salt/returners/mongo_future_return.py

```diff
@@ -141,15 +155,22 @@ def _get_conn(ret):
     # at some point we should remove support for
     # pymongo versions < 2.3 until then there are
     # a bunch of these sections that need to be supported
-
-    if float(version) > 2.3:
-        conn = pymongo.MongoClient(host, port)
+    if uri and float(version) > 2.3:
+        if uri and host:
+            warnlog = 'Specified both URI and host. URI {0} will be taken'.format(uri)
+            log.warn(warnlog)
+        pymongo.uri_parser.parse_uri(uri)
+        conn = pymongo.MongoClient(uri)
+        mdb = conn.get_database()
     else:
-        conn = pymongo.Connection(host, port)
-    mdb = conn[db_]
+        if float(version) > 2.3:
```

Comment:
would it make sense to use salt.utils.versions.LooseVersion instead of float(), to be uniform with other version checks throughout salt's codebase?

that's valid for both mongo_future_return.py and mongo_returner.py

Reply:
I can change it to the LooseVersion.

I did not want to change code not related to the URI feature but I could imagine installing salt 2018 requires lots of up to date python modules so it could be the time to remove support for pymongo < 2.3
In that case port, host and other configs could be removed and only connection string would be supported ( you can easily put all previously used options into connection string format)
If you and other salt contributors think it is OK, I can make the changes.

Reply:
I'm not that familiar with mongodb and I don't know the users' existing code-base, so I can't recommend anything. Maybe is okay to deprecate it (a sudden removal seems.. sudden).

I thought it's a low-hanging-fruit and seems among the last places where salt has a custom version compare. I think it can easily be done independent of this change though.

I'm not an official reviewer, so it's not in my power to approve things.

Reply:
Good idea, so all in all:
- depreciate host and port configuration ( but still support it)
- error on both uri and host
- any comments on removing 2.3 support anyone?

Reply:
Yes, lets use LooseVersion.

We shouldn't remove support for older versions if we can help it.  And I would say we shouldn't remove support for older configurations when it isn't absolutely necessary.